### ISSUE TITLE
Mapping to update namespace?

### ISSUE THREAD

**Issue by JesseLeite** [phpactor/phpactor#474]
In Atom I used to have a package that would let me update namespace.

For example, I could use a `class` snippet to generate a blank class, ie)

```php
<?php

class MyModel
{
    // ...
```

Then I would save the file into a PSR-4 autoloaded folder, and then hotkey to `Update Namespace`, and it would try to intelligently insert the php namespace based on the composer autoloading, ie)

```php
<?php

namespace App\Models;

class MyModel
{
    // ...
```

Similarily, if you manually moved a file into a new PSR-4 folder, and ran `Update Namespace`, it would just refactor the namespace to be correct again.

I love how you've implemented the `use` import functionality.  Think it's possible to implement a similar `update namespace` mapping?

**Comment by greg0ire** [comment]
Have you looked into `ClassNew`, and `MoveFile`

**Comment by dantleech** [comment]
- [class new](https://phpactor.github.io/phpactor/refactorings.html#class-new)
- [class move](https://phpactor.github.io/phpactor/refactorings.html#class-move)

Class new should be mapped in your `.vimrc`:

```
" Generate a new class (replacing the current file)
nmap <Leader>cc :call phpactor#ClassNew()<CR>
```

Otherwise, a "fix namespace" refactoring would be useful I think :+1:

**Comment by JesseLeite** [comment]
Yeah I've looked at ClassNew and MoveFile.  However, if I have an UltiSnip snippet setup the way I want to generate a specific class, or if I manually move anything myself (without using a Phpactor refactoring), it would still be nice to have a "fix/update namespace" refactoring option 👍

**Comment by dantleech** [comment]
Agreed this is a nice option, on some occasions I have to manually fix namespaces. Would also be nice to do this on a tree, `phpactor namespaces:fix lib/**/*.php`

**Comment by dantleech** [comment]
This can be implemented as a [Transformer](https://github.com/phpactor/code-transform/blob/master/lib/Domain/Transformer.php) using the [ClassToFIle](https://github.com/phpactor/class-to-file) (which provides a `FileToClass` converter).

**Comment by dantleech** [comment]
Made a start: https://github.com/phpactor/code-transform/pull/16/files

**Comment by JesseLeite** [comment]
❤️

**Comment by JesseLeite** [comment]
Thank you so much again for the fixing namespace & class name thing!!!

**Comment by JesseLeite** [comment]
Hmm, now that I actually go to use this, it doesn't seem to be working as [the docs suggest](https://phpactor.github.io/phpactor/refactorings.html#fix-namespace-or-class-name)?

In this example, you can see I edit the namespace (to be wrong), then call transform as the docs suggest.  But this just adds constructor, doesn't touch the namespace or class name.

![screen capture on 2018-06-18 at 13-55-20](https://user-images.githubusercontent.com/5187394/41553342-7ddd8b30-72ff-11e8-869d-060d98c53cca.gif)

I know [the docs suggest](https://phpactor.github.io/phpactor/refactorings.html#fix-namespace-or-class-name) calling `--transform=complete_constructor`, but could that have been a copy/pasta error in the docs maybe?

**Comment by dantleech** [comment]
I think you probably need to update Phpactor (the `develop` branch was only merged to `master` a few days ago). Then you should see:

![image](https://user-images.githubusercontent.com/530801/41553965-9ea3e2d6-7333-11e8-8c4c-21bae476559e.png)

Regarding the docs, they are indeed wrong, could you make a PR?

**Comment by JesseLeite** [comment]
Hmm, I actually deleted my whole `.vim/plugged` folder, ran `:PlugInstall` fresh and `composer install` in phpactor repo.  That transform option isn't visible to me.  Let me try again 🤔

**Comment by JesseLeite** [comment]
Hmm, I actually deleted my whole .vim/plugged folder, ran :PlugInstall fresh and composer install in phpactor repo. Tried manually doing a git pull as well. That transform option isn't visible to me. 🤔

**Comment by dantleech** [comment]
Hmm, I don't know if `plugged` has some sort of lock file. Maybe `:PlugUpdate` ? Also `:call phpactor#Update()`

**Comment by JesseLeite** [comment]
Ah got it.  There was permission error so the composer install was failing.  Works great now, thanks again!
